src/codegen/unused_variable.rs

```diff
@@ -50,6 +51,25 @@ pub fn should_remove_assignment(
         | Expression::Cast { expr, .. }
         | Expression::BytesCast { expr, .. } => should_remove_assignment(ns, expr, func, opt),
 
+        Expression::Builtin {
+            kind: Builtin::ArrayLength,
+            args,
+            ..
+        } => should_remove_assignment(ns, &args[0], func, opt),
+
+        Expression::Builtin {
+            kind: Builtin::ArrayPop | Builtin::ArrayPush,
+            args,
+            ..
+        } => {
+            // If the argument is a storage reference, the operation shall not be removed
+            if matches!(args[0].ty(), Type::StorageRef(..)) {
```

Comment:
Does this work on nested types?

Reply:
Do you mean user defined types?

Reply:
`Type::UserType` only supports elementary types (not vectors or structs).

`Type::Ref` is a pointer to a memory address (non-storage). `Type::StorageRef` is a storage reference. I don't think it is possible to nest one within the other.